bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+	tsf := time.Unix(ts/1000000, ts%1000000).Format("2006-01-02 15:04:05.000000")
+
+	return bigQueryMetric{
+		metricname: m.metricname,
+		tags:       string(tags),
+		value:      m.value,
+		timestamp:  tsf,
+	}, nil
+}
+
+// Upload this set of metrics to the specified BigQuery table.
+func uploadTestMetrics(w io.Writer, stream bazelStream, r *metricTestResult) error {
+	// Normalize the BigQuery table identifier based on whether one was specified
+	// in the protobuf message.
+	normalizeTableRef(&r.table)
+	fmt.Fprintf(w, "Normalized table ref: %s\n", r.table.formatTableId())
+
+	// Get client context for this BigQuery operation.
+	ctx := context.Background()
+	client, err := bq.NewClient(ctx, r.table.project)
+	if err != nil {
+		return fmt.Errorf("bigquery.NewClient: %v", err)
+	}
+	defer client.Close()
+	session := bigQuerySession{ctx: ctx, client: client}
+
+	// Check if the BigQuery dataset and table exists; create each as needed.
+	if exist := r.table.isTableExist(session); !exist {
+		ServiceStats.incrementBigQueryTableNotFound()
+		if err := createTable(w, r.table, session); err != nil {
+			return err
+		}
+	}
+
+	// Prepare the metric rows for uploading to BigQuery.
+	// Make sure each row element references its own array item.
+	// Note: bigQueryMetric implements the ValueSaver interface.
+	var bqMetrics []bigQueryMetric
+	var rows []*bigQueryMetric
+	idx := 0
+	for _, m := range r.metrics {
+		// TODO: Decide whether to override client metric time with BES server current time.
+		bqMetric, err := translateMetric(stream, m)
+		if err != nil {
+			ServiceStats.incrementBigQueryMetricDiscard(1)
+			fmt.Fprintf(w, "Discarding metric %s due to error: %s\n", m, err)
+			continue
+		}
+		bqMetrics = append(bqMetrics, bqMetric)
+		rows = append(rows, &bqMetrics[idx])
+		idx++
+	}
+	fmt.Fprintf(w, "\nTranslated metrics for BigQuery upload:\n")
+	for _, bqMetric := range bqMetrics {
+		fmt.Fprintf(w, "  %v\n", bqMetric)
+	}
+	fmt.Fprintln(w)
+
+	// Attempt to upload the metrics, assuming the dataset and table
+	// both exist. If a "not found" error occurs, sleep for a while
+	// then try uploading again.
+	//
+	// Using a finite delay loop here to give BigQuery time to instantiate
+	// the table, which can take a relativly long time (i.e. tens of seconds).
+	ok := false
+	insertionDelay := 0
+	sleepTime := 10
+	inserter := client.Dataset(r.table.dataset).Table(r.table.tablename).Inserter()
+	fmt.Fprintf(w, "Waiting for table insertion")
+	for i := 0; i < 12; i++ {
+		if err := inserter.Put(ctx, rows); err != nil {
+			// Treat anything other than a "not found" error as a failure.
+			if !strings.Contains(strings.ToLower(err.Error()), "not found") {
+				fmt.Println(w)
+				ServiceStats.incrementBigQueryInsertError()
+				return err
+			}
+			fmt.Fprintf(w, ".")
+		} else {
+			ok = true
+			break
+		}
+		time.Sleep(time.Duration(sleepTime) * time.Second)
+		insertionDelay += sleepTime
+	}
+	fmt.Fprintln(w)
+	if !ok {
```

Comment:
It is probably better to start a goroutine for incrementing the prometheus stats. Just so that it's non blocking.

Reply:
so just call for example `go ServiceStats.incrementBigQueryMetricUpload(len(rows))`

Reply:
OK, will do.